src/luv.c

```diff
@@ -506,20 +512,50 @@ static void luv_handle_init(lua_State* L) {
   lua_setfield(L, LUA_REGISTRYINDEX, "uv_stream");
 }
 
-LUALIB_API lua_State* luv_state(uv_loop_t* loop) {
-  return (lua_State*)loop->data;
+static const char* luv_loop_key = "uv_loop";
+static const char* luv_state_key = "main_thread";
+
+LUALIB_API lua_State* luv_state(lua_State* L) {
+  lua_State* lstate;
+  lua_pushlightuserdata(L, (void*)luv_state_key);
+  lua_rawget(L, LUA_REGISTRYINDEX);
+  lstate = lua_tothread(L, -1);
+  lua_pop(L, 1);
+  if (lstate==NULL) {
+    luaL_error(L, "cannot get main thread");
+  }
+  return lstate;
 }
 
-// TODO: find out if storing this somehow in an upvalue is faster
 LUALIB_API uv_loop_t* luv_loop(lua_State* L) {
   uv_loop_t* loop;
-  lua_pushstring(L, "uv_loop");
+  lua_pushlightuserdata(L, (void*)luv_loop_key);
```

Comment:
Maybe keep this as `pushstring` to avoid light userdata? Same with `luv_state_key`.

Reply:
Ok. Is there a more quick way? those api are repeat loop a large numbers times.

Reply:
Is there a reason it needs to be retrieved from the Lua registry each time? Could we store a pointer to the loop in C as well and use that?

Reply:
Not each time,  mainly when luv_setup_req or luv_setup_handle,